cmake/NeuronTestHelper.cmake

```diff
@@ -0,0 +1,361 @@
+  set(binary_target_name "${prefix}_${nrnivmodl_command_hash}")
+  set(nrnivmodl_working_directory "${group_working_directory}/${nrnivmodl_command_hash}")
+  # Short-circuit in case we set up these rules already
+  if(NOT TARGET ${binary_target_name})
+    # Construct the names of the modfiles in the build tree, i.e. the filenames from ${modfiles}
+    # with the path ${nrnivmodl_working_directory} in front
+    foreach(modfile ${modfiles})
+      get_filename_component(modfile_name "${modfile}" NAME)
+      list(APPEND modfile_build_paths "${nrnivmodl_working_directory}/${modfile_name}")
+    endforeach()
+    # Add a custom command that copies the modfiles into the build tree, the nrnivmodl command can
+    # then depend on its input modfiles. copy_if_different would create the directory if we had >1
+    # modfile, but in case there is only one then we should create the directory manually.
+    file(MAKE_DIRECTORY "${nrnivmodl_working_directory}")
+    add_custom_command(
+      OUTPUT ${modfile_build_paths}
+      DEPENDS ${modfiles}
+      COMMAND ${CMAKE_COMMAND} -E copy_if_different ${modfiles} "${nrnivmodl_working_directory}"
+      COMMENT "Copying modfiles needed for test ${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}")
+    # Construct the names of the important output files
+    set(special "${nrnivmodl_working_directory}/${CMAKE_HOST_SYSTEM_PROCESSOR}/special")
+    # Add the custom command to generate the binaries. Get nrnivmodl from the build directory. At
+    # the moment it seems that `nrnivmodl` is generated at configure time, so there is no target to
+    # depend on and it should always be available, but it will try and link against libnrniv.so and
+    # libcoreneuron.so so we must depend on those. TODO: could the logic of `nrnivmodl` be
+    # translated to CMake, so it can be called natively here and the `nrnivmodl` executable would be
+    # a wrapper that invokes CMake?
+    set(output_binaries "${special}")
+    set(nrnivmodl_dependencies nrniv_lib)
+    if(requires_coreneuron)
+      list(APPEND output_binaries "${special}-core")
+      list(APPEND nrnivmodl_dependencies coreneuron)
+    endif()
+    add_custom_command(
+      OUTPUT ${output_binaries}
+      DEPENDS ${nrnivmodl_dependencies} ${modfile_build_paths}
+      COMMAND ${CMAKE_COMMAND} -E env ${NRN_TEST_ENV} PATH=${CMAKE_BINARY_DIR}/bin:$ENV{PATH}
+              ${nrnivmodl_command}
+      COMMENT "Building special[-core] for test ${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}"
+      WORKING_DIRECTORY ${nrnivmodl_working_directory})
+    # Add a target that depends on the binaries
+    add_custom_target(${binary_target_name} DEPENDS ${output_binaries})
+    # Make the test-group-level target depend on this new target.
+    add_dependencies(${prefix} ${binary_target_name})
+  endif()
+  # Set up the actual test. First, collect the script files that need to be copied into the test-
+  # specific working directory and copy them there.
```

Comment:
I believe copying scripts around won't work for real world tests. Often we have some input data, other configs files and they expect we are executing everything from the `pwd`. So either we need to disable this or have some option to say this test can't be run from different directory.

Reply:
I don't see a **fundamental** problem here, you "just" have to specify more patterns and copy all the relevant files across; it doesn't matter that some of them are config files and others are scripts. If it gets too unwieldy, maybe we can just make sure that `*` works as expected.

I was aiming for `make test` to run entirely from the build directory (i.e. not refer to the source directory at all), and to ensure that dependencies are declared properly so `make` can do the right thing (i.e. `{build_dir}/path/to/script.py` gets updated by `make` if `{source_dir}/path/to/script.py` is changed).

Reply:
> I don't see a fundamental problem here, you "just" have to specify more patterns and copy all the relevant files across; it doesn't matter that some of them are config files and others are scripts. If it gets too unwieldy, maybe we can just make sure that * works as expected.

Agree! 👍 

> I was aiming for make test to run entirely from the build directory

This would be good to have indeed! (and also for parallel runs)